cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll

```diff
@@ -255,6 +255,36 @@ class OperatorNewAllocationFunction extends AllocationFunction {
   }
 }
 
+/**
+ * The predicate analyzes a `sizeExpr`, which is an argument to an allocation
+ * function like malloc, and tries to split it into an expression `lengthExpr`
+ * that describes the length of the allocated array, and the size of the allocated
+ * element type `sizeof`.
+ * If this is not possible, the allocation is considered to be of size 1 and of
+ * length `sizeExpr`.
+ */
+private predicate deconstructSizeExpr(Expr sizeExpr, Expr lengthExpr, int sizeof) {
+  if
+    sizeExpr instanceof MulExpr and
+    exists(SizeofOperator sizeofOp, Expr lengthOp |
+      sizeofOp = sizeExpr.(MulExpr).getAnOperand() and
+      lengthOp = sizeExpr.(MulExpr).getAnOperand() and
+      not lengthOp instanceof SizeofOperator and
+      exists(sizeofOp.getValue().toInt())
+    )
+  then
+    exists(SizeofOperator sizeofOp |
+      sizeofOp = sizeExpr.(MulExpr).getAnOperand() and
+      lengthExpr = sizeExpr.(MulExpr).getAnOperand() and
+      not lengthExpr instanceof SizeofOperator and
+      sizeof = sizeofOp.getValue().toInt()
+    )
```

Comment:
Why is it important that we're multiplying by a `sizeof` rather than a constant value (anything with a `getValue` result) that was obtained by other means? (This is not a question that needs to block this PR).

Reply:
Two reasons:
1. Using `sizeof`, the user clearly states their intent as using a size multiplier (the predicate is named `getSizeMult`, so it shouldn't return values that are not clearly sizes of the allocation.
This could go even further by using the `getAllocatedElementType` from #3337, and ensuring that either `getSizeMult()=1` or ´getSizeMult() = getAllocatedElementType().getSize() = getSizeMult` holds.
2. I don't think  we want that the predicate has multiple results, so in the cases `malloc(5*sizeof(int))` and `malloc(5*3)` we'd need a canonical way to only report one size multiplier (else predicates like `getSizeBytes` would also evaluate to multiple values)

Reply:
That sounds reasonable. We can always adapt to coding patterns we observe later.